Remove six

hramezani opened the pull request:
<!---
Thanks for your contribution! ♥️

If this is your first PR to urllib3 please review the Contributing Guide:
https://urllib3.readthedocs.io/en/latest/contributing.html

Adhering to the Contributing Guide means we can review, merge, and release your change faster! :)
--->

There is another possibility to remove the dependency to six.
we can move the required function of six like `ensure_str`, `ensure_binary`, ... and then remove `six.py`

sethmlarson commented on src/urllib3/util/request.py:
```diff
@@ -80,11 +79,13 @@ def make_headers(
 
     if basic_auth:
-        headers["authorization"] = "Basic " + b64encode(b(basic_auth)).decode("utf-8")
+        headers["authorization"] = "Basic " + b64encode(
```
`b64encode` can only return ASCII so we can do `b64encode(x.encode("latin-1")).decode()`. Also we can use an `f-string here instead of string concat

sethmlarson commented on test/test_filepost.py:
```diff
@@ -2,9 +2,9 @@
-from urllib3.packages.six import b
 
 BOUNDARY = "!! test boundary !!"
+LATIN_ENCODED_BOUNDARY = BOUNDARY.encode("latin-1")
```
Instead of `LATIN_ENCODED_BOUNDARY` maybe `BOUNDARY_BYTES` is a more descriptive name?

sethmlarson commented on test/test_response.py:
```diff
@@ -801,7 +800,7 @@ def test_invalid_chunk_length(self):
         orig_ex = ctx.value.args[1]
         assert isinstance(orig_ex, InvalidChunkLength)
-        assert orig_ex.length == six.b(fp.BAD_LENGTH_LINE)
+        assert orig_ex.length == fp.BAD_LENGTH_LINE.encode("latin-1")
```
We should remove it and only do `.encode()` to not confuse future readers into thinking there's anything relevant about the "latin-1" encoding here.

sethmlarson commented on test/with_dummyserver/test_connectionpool.py:
```diff
@@ -700,7 +699,7 @@ def test_source_address(self):
             ) as pool:
                 r = pool.request("GET", "/source_address")
-                assert r.data == b(addr[0])
+                assert r.data == addr[0].encode("latin-1")
```
Same as above, remove latin-1 encoding.

sethmlarson reviewed (changes requested):
Thanks for this! I left you some comments.

hramezani commented:
@sethmlarson Thanks for the review. I addressed your comments.

What is your idea about my comment:

> we can move the required function of six like ensure_str, ensure_binary, ... and then remove six.py

If you agree with moving these function out of `six.py` please give me some hint about the file that we need to move these function to it.

sethmlarson commented:
Definitely, maybe create a new module (`urllib3.util.util`??) since `urllib3.util` is unfortunately public.

hramezani commented:
> Definitely, maybe create a new module (`urllib3.util.util`??) since `urllib3.util` is unfortunately public.

I will try to make a PR for it.

sethmlarson commented:
Okay, so this PR is good to go then?

hramezani commented:
we can merge this one and make another one for the six functions. Or have them all in this patch.
Which one do you prefer @sethmlarson?

sethmlarson commented:
Let's bring everything into this PR, then we can label it "remove six"

sethmlarson commented on test/test_filepost.py:
```diff
@@ -2,9 +2,9 @@
-from urllib3.packages.six import b
 
 BOUNDARY = "!! test boundary !!"
+BOUNDARY_BYTES = BOUNDARY.encode("latin-1")
```
Can we remove the latin-1 here as well?

sethmlarson commented on src/urllib3/util/util.py:
```diff
@@ -0,0 +1,36 @@
+def ensure_binary(s, encoding="utf-8", errors="strict"):
```
Thoughts on using `to_str` and `to_bytes` as both shorter and more descriptive?

The following are optimized such that `to_X()` on type `X` is fast and we only pay the "not isinstance()" tax when it's not `X`. Also calling without any arguments is a lot faster than with any arguments so we should optimize for this case:

```python
from typing import Union, Optional


def to_bytes(x: Union[str, bytes], encoding: Optional[str] = None, errors: Optional[str] = None) -> bytes:
    if isinstance(x, bytes):
        return x
    elif not isinstance(x, str):
        raise TypeError(f"not expecting type {type(x).__name__}")
    elif encoding or errors:
        return x.encode(encoding, errors=errors)
    return x.encode()


def to_str(x: Union[str, bytes], encoding: Optional[str] = None, errors: Optional[str] = None) -> str:
    if isinstance(x, str):
        return x
    elif not isinstance(x, bytes):
        raise TypeError(f"not expecting type {type(x).__name__}")
    elif encoding or errors:
        return x.decode(encoding, errors=errors)
    return x.decode()
```

That way `to_bytes()` will encode as ASCII with errors=strict

hramezani replied:
@sethmlarson Thanks for the suggestion.
I would suggest using default values for `encoding` and `errors` because with 
```python
    elif encoding or errors:
        return x.decode(encoding, errors=errors)
````
If we pass just `encoding`, the condition will be True and we will face `TypeError` (because the `error` is None)
`TypeError: decode() argument 2 must be str, not None`

sethmlarson replied:
Ah yes, perhaps `x.encode(encoding or "ascii", errors=errors or "strict")`. Trying to detect the case when we can simple `x.decode()` since it's >4x faster than calling with any arguments and encoding/decoding is a common operation for us.

sethmlarson commented on src/urllib3/util/url.py:
```diff
@@ -210,7 +210,7 @@ def _encode_invalid_chars(component, allowed_chars, encoding="utf-8"):
         return component
 
-    component = six.ensure_text(component)
+    component = ensure_str(component)
```
This call to `to_str()` would need an `encoding="utf-8"` if implemented like below.

hramezani replied:
> would need an encoding="utf-8" if implemented like below.

Do you mean `encoding=encoding`? because the `_encode_invalid_chars` accepts an `encoding` param with default value to `"utf-8"`

sethmlarson replied:
Hm, that parameter has been doing basically nothing it seems. You can remove the parameter and at the end of the `_encode_invalid_chars()` change `encoded_component.decode(encoding)` to `.decode()` since percent-encoding implies the end result is ASCII.

sethmlarson commented on src/urllib3/util/url.py:
```diff
@@ -396,7 +396,7 @@ def parse_url(url):
 
     def ensure_type(x):
-        return x if x is None else six.ensure_text(x)
+        return x if x is None else ensure_str(x)
```
Idle thought: I feel like this isn't necessary anymore since we only accept URLs that are `str` for Python 3 so URL components can only end up as `str`.

sethmlarson reviewed (commented):
A suggestion on name/implementation, would like your thoughts

sethmlarson commented on src/urllib3/contrib/securetransport.py:
```diff
@@ -869,7 +867,7 @@ def set_alpn_protocols(self, protocols):
                 "SecureTransport supports ALPN only in macOS 10.12+"
             )
-        self._alpn_protocols = [six.ensure_binary(p) for p in protocols]
+        self._alpn_protocols = [util.util.to_bytes(p) for p in protocols]
```
Let's expand the import to `from ..util.util import to_bytes` so it's similar to the rest of the usages

hramezani replied:
I get this error when I import it by `from ..util.util import to_bytes`

`Could not import SecureTransport: ModuleNotFoundError("No module named 'urllib3.packages.util'")`

sethmlarson replied:
`to_bytes(p, "ascii")`

sethmlarson commented on src/urllib3/contrib/pyopenssl.py:
```diff
@@ -450,7 +449,7 @@ def load_cert_chain(self, certfile, keyfile=None, password=None):
 
     def set_alpn_protocols(self, protocols):
-        protocols = [six.ensure_binary(p) for p in protocols]
+        protocols = [util.util.to_bytes(p) for p in protocols]
```
nit: Let's expand the import to `from ..util.util import to_bytes` so it's similar to the rest of the usages

hramezani replied:
doc tests will fail with this change.

Maybe it is related to `from .. import util` that exists at the top of file

sethmlarson replied:
This should be `to_bytes(p, "ascii")`

sethmlarson commented on src/urllib3/util/url.py:
```diff
@@ -395,17 +395,14 @@ def parse_url(url):
-        scheme=ensure_type(scheme),
-        auth=ensure_type(auth),
-        host=ensure_type(host),
+        scheme=scheme,
```
:tada: This is so great!

sethmlarson commented on src/urllib3/util/util.py:
```diff
@@ -0,0 +1,37 @@
+from typing import Optional, Union
+
+
+def to_bytes(
```
Since I wrote this code would be good to get another outside review (cc @pquentin @hodbn )

sethmlarson reviewed (dismissed):
besides the two nits and the call for reviews I'm happy with this change, goodbye six!

sethmlarson commented:
Oh also I wonder if there's some coverage loss because the `to_X()` `TypeError` case isn't covered. Maybe can throw in a couple test cases for our new utility methods? It's fine to mark `reraise` as not requiring code coverage since we lifted it from six.

hramezani commented:
@sethmlarson Thanks for the review.

I added two tests to cover `TypeError` in the newly adde to `to_X()` functions.
Also, I ignored coverage for `def reraise` by adding it to the `exclude_lines ` section of `.coveragerc` (not sure that I did it right)

pquentin reviewed (changes requested):
Great work! I have a question about `to_bytes`

pquentin commented:
Also, to_bytes and to_str are not fully covered as you can see here: https://codecov.io/gh/urllib3/urllib3/src/3d79bccbda1486de6945d0a084b782a43c77890e/src/urllib3/util/util.py

<img width="857" alt="Capture d’écran 2020-11-22 à 21 17 36" src="https://user-images.githubusercontent.com/42327/99910561-27768880-2d08-11eb-863e-ddc32a931a07.png">

hramezani commented on src/urllib3/util/util.py:
```diff
@@ -0,0 +1,37 @@
+        return x
+    elif not isinstance(x, str):
+        raise TypeError(f"not expecting type {type(x).__name__}")
+    elif encoding or errors:
+        return x.encode(encoding or "ascii", errors=errors or "strict")
+    return x.encode()
```
Since we want to call `x.encode()`, because it's faster (based on @sethmlarson [comment](https://github.com/urllib3/urllib3/pull/2078#discussion_r527957964). I think we can change it like this:

`return x.encode(encoding or "utf-8", errors=errors or "strict")`

sethmlarson replied:
Wow, I didn't realize the default encoding for `str.encode()` was utf-8, thought it was ascii all this time. Thanks for bringing this to my attention!

In this case we will need to explicitly say `to_str("ascii")` in all the places I previously thought should be ASCII.

hramezani commented:
> Also, to_bytes and to_str are not fully covered as you can see here: https://codecov.io/gh/urllib3/urllib3/src/3d79bccbda1486de6945d0a084b782a43c77890e/src/urllib3/util/util.py
> 
> <img alt="Capture d’écran 2020-11-22 à 21 17 36" width="857" src="https://user-images.githubusercontent.com/42327/99910561-27768880-2d08-11eb-863e-ddc32a931a07.png">

I added some tests to complete the coverage.

sethmlarson commented on src/urllib3/util/url.py:
```diff
@@ -203,14 +203,14 @@ def split_first(s, delims):
         return component
 
-    component = six.ensure_text(component)
+    component = to_str(component, encoding="utf-8")
```
Can remove the `encoding=` here

sethmlarson commented on src/urllib3/util/url.py:
```diff
@@ -286,7 +286,7 @@ def _normalize_host(host, scheme):
                     return host.lower()
             elif not IPV4_RE.match(host):
-                return six.ensure_str(
+                return to_str(
```
`to_str(x, "ascii")`

sethmlarson reviewed (changes requested):
Sorry for the confusion, thank you @pquentin for correcting me. A few little comments left!

sethmlarson reviewed (dismissed):
This looks good to me, looking for a final review from @pquentin now :)

pquentin commented on test/test_util.py:
```diff
@@ -776,6 +777,41 @@ def test_create_connection_with_valid_idna_labels(self, socket, getaddrinfo, hos
+            ("test", {}, "test"),  # str input
+            (b"test", {}, "test"),  # bytes input
+            (b"test", {"encoding": "utf-8"}, "test"),  # bytes input with utf-8
+            (b"test", {"encoding": "ascii"}, "test"),  # bytes input with asscii
```
```suggestion
            (b"test", {"encoding": "ascii"}, "test"),  # bytes input with ascii
```

pquentin commented on test/test_util.py:
```diff
@@ -776,6 +777,41 @@ def test_create_connection_with_valid_idna_labels(self, socket, getaddrinfo, hos
+        assert to_str(input, **params) == expected
+
+    def test_to_str_error(self):
+        with pytest.raises(TypeError) as e:
+            to_str(1)
+        assert "not expecting type int" in str(e.value)
```
```suggestion
        with pytest.raises(TypeError, match="not expecting type int"):
            to_str(1)
```

pquentin commented on test/test_util.py:
```diff
@@ -776,6 +777,41 @@ def test_create_connection_with_valid_idna_labels(self, socket, getaddrinfo, hos
+        assert to_bytes(input, **params) == expected
+
+    def test_to_bytes_error(self):
+        with pytest.raises(TypeError) as e:
+            to_bytes(1)
+        assert "not expecting type int" in str(e.value)
```
```suggestion
        with pytest.raises(TypeError, match="not expecting type int"):
            to_bytes(1)
```

pquentin commented on .coveragerc:
```diff
@@ -22,6 +22,7 @@ exclude_lines =
     pass
     import
     raise
+    def reraise
```
This file is for statements that occur multiple times, which is not the case here. And `reraise` is covered anyway, so we can remove that line.

```suggestion
```

pquentin commented on src/urllib3/util/util.py:
```diff
@@ -0,0 +1,37 @@
+        return x
+    elif not isinstance(x, str):
+        raise TypeError(f"not expecting type {type(x).__name__}")
+    elif encoding or errors:
+        return x.encode(encoding or "utf-8", errors=errors or "strict")
+    return x.encode()
```
You can remove one branch here:

```suggestion
    return x.encode(encoding or "utf-8", errors=errors or "strict")
```

hramezani replied:
I think @sethmlarson suggested the branch in function to detect the case that we can call `x.encode()` because it's 4X faster. 
Take a look at this [comment](https://github.com/urllib3/urllib3/pull/2078#discussion_r527972377)

pquentin commented on src/urllib3/util/util.py:
```diff
@@ -0,0 +1,37 @@
+        return x
+    elif not isinstance(x, bytes):
+        raise TypeError(f"not expecting type {type(x).__name__}")
+    elif encoding or errors:
+        return x.decode(encoding or "utf-8", errors=errors or "strict")
+    return x.decode()
```
You can remove one branch here:

```suggestion
    return x.decode(encoding or "utf-8", errors=errors or "strict")
```

pquentin commented on test/test_util.py:
```diff
@@ -776,6 +777,41 @@ def test_create_connection_with_valid_idna_labels(self, socket, getaddrinfo, hos
+            ("test", {}, b"test"),  # bytes input
+            ("é", {}, b"\xc3\xa9"),  # bytes input
+            ("test", {"encoding": "utf-8"}, b"test"),  # bytes input with utf-8
+            ("test", {"encoding": "ascii"}, b"test"),  # bytes input with asscii
```
```suggestion
            ("test", {"encoding": "ascii"}, b"test"),  # bytes input with ascii
```

pquentin reviewed (changes requested):
Great work! We're nearly there, I only have cosmetic comments left.

pquentin commented:
This now looks good to *me*, but contains code that I wrote, so I need review from either or you now!

pquentin commented:
Weird, but I seem to reproduce. I guess a comment would be best. I removed that part from the last commit, please review

sethmlarson reviewed (approved):
@pquentin LGTM, merge when ready!

pquentin commented:
Closing/reopening as we ran into the coverage SQLite race condition again.

pquentin commented:
Goodbye six 👋 You served us well

hramezani commented:
@sethmlarson @pquentin thanks for the review and your help.
